### DISCUSSION TITLE
Access to an undefined property App\Models\User::$active

### DISCUSSION THREAD

**Discussion by faridsa** [larastan/larastan#1767] [Q&A]
At this point this error is not understandable for me

My User model:
```php
namespace App\Models;

use Illuminate\Contracts\Auth\Access\Authorizable;
use Illuminate\Contracts\Auth\MustVerifyEmail;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Notifications\Notifiable;
use Laravel\Sanctum\HasApiTokens;

class User extends Authenticatable implements MustVerifyEmail
{
    use HasApiTokens, HasFactory, Notifiable;

    /**
     * The attributes that are mass assignable.
     *
     * @var string[]
     */
    protected $fillable = [
        'name',
        'surname',
        'email',
        'email_verified_at',
        'password',
        'password_changed_at',
        'active',
        'last_login_at',
        'last_login_ip',
        'remember_token',
        'type',
    ];


    /**
     * The attributes that should be hidden for serialization.
     *
     * @var array<int, string>
     */
    protected $hidden = [
        'password',
        'remember_token',
    ];


    /**
     * @var array<string, string>
     */
    protected $casts = [
        'email_verified_at' => 'datetime',
        'last_login_at' => 'datetime',
    ];


    public function isActive(): bool
    {
        return $this->active ? true : false;
    }

    public function isVerified(): bool
    {
        return filled($this->email_verified_at) ? true : false;
    }

    public function userProfile()
    {
        return $this->hasOne(\App\Models\UserProfile::class);
    }
}
```

I have some others errors with models in the same situation. Larastan reports undefined properties in models which have those properties defined, so maybe solving this situation I hope I could solve others.

**Comment by mad-briller** [comment]
larastan is [reading your migrations](https://github.com/nunomaduro/larastan/blob/2.x/docs/custom-config-parameters.md#databasemigrationspath) to understand what properties are available on the models, does larastan have access to all of your migrations? are those columns added by a migration or by some other means?

**Reply by faridsa** [reply]
Thank you for your reply. 
I think that there is nothing blocking access from larastan to my migrations.
So the `$active` property which is firing larastan error is defined both in `$fillable` array and table migration

My user migration file is:
```
<?php

use Illuminate\Database\Migrations\Migration;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;

class CreateUsersTable extends Migration
{
    /**
     * Run the migrations.
     *
     * @return void
     */
    public function up()
    {
        Schema::create('users', function (Blueprint $table) {
            $table->id();
            $table->string('name');
            $table->string('surname');
            $table->string('email')->unique();
            $table->string('password');
            $table->string('type');
            $table->string('last_login_ip');
            $table->boolean('active')->nullable();
            $table->timestamp('email_verified_at')->nullable();
            $table->timestamp('password_changed_at')->nullable();
            $table->timestamp('last_login_at')->nullable();
            $table->rememberToken();
            $table->timestamps();
        });
    }

    /**
     * Reverse the migrations.
     *
     * @return void
     */
    public function down()
    {
        Schema::dropIfExists('users');
    }
}
```

While my phpstan.neon file is:

```
includes:
    - ./vendor/nunomaduro/larastan/extension.neon

parameters:

    paths:
        - app/

    # Level 9 is the highest level
    level: 5

#    ignoreErrors:
#        - '#PHPDoc tag @var#'
#
#    excludePaths:
#        - ./*/*/FileToBeExcluded.php
#
#    checkMissingIterableValueType: false
```

**Reply by mad-briller** [reply]
judging by the migration that should work fine, do the other custom properties other than `active` such as `type` of the `User` model work correctly?

**Reply by mad-briller** [reply]
what happens if you add `active` to the `$casts` property?

**Reply by faridsa** [reply]
All other properties in this model are working fine according to larastan. It only reports error in $active property.

I've tried also adding to $cast array but with no luck. Error remains the same.

![Screenshot 2023-10-17 at 12 51 38](https://github.com/nunomaduro/larastan/assets/23212965/dcc04a05-027b-4030-81f3-4b733dc81018)

**Reply by szepeviktor** [reply]
I see!!
You access `active` inside the model.
Does this error occur *outside* the model?

**Reply by faridsa** [reply]
Hi Viktor, in fact I have no errors in production, it just appears when running larastan, but any call to $user->isActive() returns correctly true or false.

So, I'm trying basically to understand the nature of this error, but I'm not having any error or warning in logs about it.

**Reply by szepeviktor** [reply]
Does Larastan report you an error with `active` outside the `User.php`?

**Reply by faridsa** [reply]
If you are referring to any call from a controller to verify this property, no, there is no other error report about $active.

**Reply by szepeviktor** [reply]
Thank you!
I think Larastan does not yet know about that property in the model.
@mad-briller @canvural Please comment on this.

**Comment by rabol** [comment]
Hi

I have more or less the same problem with a property, it's really strange that one property is reported as not existing ip_address, several other properties are not reported.

adding a phpdoc block

```
/** @property string $ip_address */
```

solves the issue, but I would also like to understand why it only one of the properties on the model is an issue
